Use imaging mode generated code only

pchintalapudi opened the pull request:
This should remove the last codepaths within codegen that affect the generated LLVM IR based on imaging mode or not.

~The majority of the remaining uses of imaging_mode are now in creating modules, where it's used to decide if we emit with DWARFv2 or DWARFv4, which I think we can move to just aotcompile?~ That's gone now, thanks @gbaraldi.

pchintalapudi committed:
Remove imaging mode from most of codegen

gbaraldi commented:
Have we tried just using dwarfv4 everywhere? it seems that what clang uses on macos?

gbaraldi committed:
Remove imaging mode option from LLVM module creation (#50853)

According to clang macos supports dwarfv4 from above macos 10.10. Julia
requires 10.14 so we can remove the branch.
Ref
https://github.com/llvm/llvm-project/blob/f62c9252fc0f1fa0a0f02033659db052c2202a4c/clang/lib/Driver/ToolChains/Darwin.cpp#L1243

vtjnash reviewed (commented):
SGTM, though now needs a rebase

vtjnash commented on src/cgutils.cpp:
```diff
@@ -345,7 +345,7 @@ static Constant *julia_pgv(jl_codectx_t &ctx, const char *cname, void *addr)
     std::string gvname;
     if (!gv) {
-        uint64_t id = ctx.emission_context.imaging_mode ? jl_atomic_fetch_add(&globalUniqueGeneratedNames, 1) : ctx.emission_context.global_targets.size();
+        uint64_t id = jl_atomic_fetch_add(&globalUniqueGeneratedNames, 1);
```
Once we eventually fix the behavior of our linker pass, this should be reverted to the other side of the branch
```suggestion
        uint64_t id = jl_atomic_fetch_add(&globalUniqueGeneratedNames, 1); // TODO: use ctx.emission_context.global_targets.size()
```

pchintalapudi committed:
Apply suggestions from code review

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
